libraries/pod/src/option.rs

```diff
@@ -0,0 +1,138 @@
+//! Generic `Option` that can be used as a `Pod` for types that can have
+//! a designated `None` value.
+//!
+//! For example, a 64-bit unsigned integer can designate `0` as a `None` value.
+//! This would be equivalent to
+//! [`Option<NonZeroU64>`](https://doc.rust-lang.org/std/num/type.NonZeroU64.html)
+//! and provide the same memory layout optimization.
+
+use {
+    bytemuck::{Pod, Zeroable},
+    solana_program::{program_option::COption, pubkey::Pubkey},
+};
+
+/// Trait for types that can be `None`.
+///
+/// This trait is used to indicate that a type can be `None` according to a
+/// specific value.
+pub trait Nullable: Default + Pod {
+    /// Indicates whether the value is `None` or not.
+    fn is_none(&self) -> bool;
+
+    /// Indicates whether the value is `Some`` value of type `T`` or not.
+    fn is_some(&self) -> bool {
+        !self.is_none()
+    }
+}
+
+/// A "pod-enabled" type that can be used as an `Option<T>` without
+/// requiring extra space to indicate if the value is `Some` or `None`.
+///
+/// This can be used when a specific value of `T` indicates that its
+/// value is `None`.
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct PodOption<T: Nullable>(T);
+
+impl<T: Nullable> PodOption<T> {
+    /// Returns the contained value as an `Option`.
+    #[inline]
+    pub fn get(self) -> Option<T> {
+        if self.0.is_none() {
+            None
+        } else {
+            Some(self.0)
+        }
+    }
+
+    /// Returns the contained value as an `Option`.
+    #[inline]
+    pub fn as_ref(&self) -> Option<&T> {
+        if self.0.is_none() {
+            None
+        } else {
+            Some(&self.0)
+        }
+    }
+
+    /// Returns the contained value as a mutable `Option`.
+    #[inline]
+    pub fn as_mut(&mut self) -> Option<&mut T> {
+        if self.0.is_none() {
+            None
+        } else {
+            Some(&mut self.0)
+        }
+    }
+}
+
+unsafe impl<T: Nullable> Pod for PodOption<T> {}
+
+unsafe impl<T: Nullable> Zeroable for PodOption<T> {}
+
+impl<T: Nullable> From<T> for PodOption<T> {
+    fn from(value: T) -> Self {
+        PodOption(value)
+    }
+}
+
+impl<T: Nullable> From<Option<T>> for PodOption<T> {
+    fn from(from: Option<T>) -> Self {
+        match from {
+            Some(value) => PodOption(value),
+            None => PodOption(T::default()),
```

Comment:
This assumes that `T::default()` is the `None` value, which is not guaranteed by the trait. I understand the desire to make the value flexible, so to make that behavior less error-prone, how about changing the trait to something like:

```
pub trait Nullable: Pod + PartialEq + Sized {
    /// Value indicating the `None` to check against
    const NONE: Self; 
    /// Indicates whether the value is `None` or not.
    fn is_none(&self) -> bool {
        self == &Self::NONE
    }
```

Reply:
The only caveat is that in this case `NONE` value has to be a `const`. I don't think is a big deal, but perhaps we can consider adding a method to the trait to specify the `NONE` value.

* Using `const NONE`:
  ```rust
  impl Nullable for Pubkey {
      const NONE: Self = Pubkey::new_from_array([0u8; PUBKEY_BYTES]);
  }
  ```
* Using `none()` method:
  ```rust
  impl Nullable for Pubkey {
      fn none() -> Self {
          Pubkey::default()
      }
  }
  ```